### PR TITLE
Allow .proto files to import between Haskell packages.

### PR DESCRIPTION
**Pull request by judah**
Also adds the `proto-lens-protobuf-types` package which includes
`Proto.Google.Protobuf.{Any,Duration,Wrappers}`.

This change allows one Haskell package to generate bindings for `foo.proto`,
and another package to generate bindings for `bar.proto`, where
`bar.proto` imports `foo.proto`.

Under the hood, each package copies its `.proto` files into
`$datadir/proto-lens-imports`, so that other packages can find
and use them when compiling their own `.proto` files.

Note: a minor corner case is not handled correctly: if a test or
benchmark requires va .proto that imports from another package,
the *library* also needs to specify that dependency.  See
`proto-lens-tests.cabal` for more details.

### CODE REVIEW

**Comment by judah** [comment]
@awpr, interested in taking a look?

File: proto-lens-protoc/src/Data/ProtoLens/Setup.hs [hunk 1]
```diff
@@ -49,9 +63,16 @@ import System.FilePath
...
+    )
+import System.Directory
+    ( createDirectoryIfMissing
+    , doesDirectoryExist
+    , findExecutable
+    , removeDirectoryRecursive
     )
-import System.Directory (createDirectoryIfMissing, findExecutable)
 import System.Process (callProcess)
+import Distribution.Verbosity (Verbosity)
```
Comment (blackgnezdo) [inline proto-lens-protoc/src/Data/ProtoLens/Setup.hs]: Sort?

File: proto-lens-protoc/src/Data/ProtoLens/Setup.hs [hunk 2]
```diff
@@ -123,21 +144,67 @@ generatingSpecificProtos
...
+    { buildHook = \p l h f -> generate p l >> buildHook hooks p l h f
+    , haddockHook = \p l h f -> generate p l >> haddockHook hooks p l h f
+    , replHook = \p l h f args -> generate p l >> replHook hooks p l h f args
     , sDistHook = \p maybe_l h f -> case maybe_l of
             Nothing -> error "Can't run protoc; run 'cabal configure' first."
             Just l -> do
-                        generateSources p l
+                        generate p l
                         sDistHook hooks (fudgePackageDesc l p) maybe_l h f
+    , postCopy = \a flags pkg lbi -> do
```
Comment (blackgnezdo) [inline proto-lens-protoc/src/Data/ProtoLens/Setup.hs]: Out of curiosity, why this style over
postCopy a flags pkg lbi = do
?

File: proto-lens-protoc/src/Data/ProtoLens/Setup.hs [hunk 3]
```diff
@@ -123,21 +144,67 @@ generatingSpecificProtos
...
+generateSources :: FilePath -- ^ The root directory
+                -> LocalBuildInfo
+                -> [FilePath] -- ^ Proto files relative to the root directory.
+                -> IO ()
+generateSources root l files = do
+    -- Collect import paths from build-depends of this package.
+    importDirs <- filterM doesDirectoryExist
+                     [ InstalledPackageInfo.dataDir info </> protoLensImportsPrefix
+                     | (_, c, _) <- componentsConfigs l
+                     , (_, i ) <- componentPackageDeps c
```
Comment (blackgnezdo) [inline proto-lens-protoc/src/Data/ProtoLens/Setup.hs]: Remove space in i) ?

File: proto-lens-protoc/src/Data/ProtoLens/Setup.hs (hunk 1 above)
Comment (judah) [inline proto-lens-protoc/src/Data/ProtoLens/Setup.hs]: Done.

File: proto-lens-protoc/src/Data/ProtoLens/Setup.hs (hunk 2 above)
Comment (judah) [inline proto-lens-protoc/src/Data/ProtoLens/Setup.hs]: Haskell's syntax doesn't let you do that when defining fields in a record :(

File: proto-lens-protoc/src/Data/ProtoLens/Setup.hs (hunk 3 above)
Comment (judah) [inline proto-lens-protoc/src/Data/ProtoLens/Setup.hs]: Done.
